reconcile/acs_rbac.py

```diff
@@ -0,0 +1,562 @@
+
+            # role portion
+            # access scope is included in diff check once more here
+            # in case the role needs to be assigned different access scope.
+            # changes to access scope resource are handled in dedicated section above
+            # assignments are not included in this diff. Handled in dedicated section above
+            if role_diff_pair.current.diff_role(role_diff_pair.desired):
+                if not dry_run:
+                    try:
+                        acs.update_role(
+                            role_diff_pair.desired.name,
+                            role_diff_pair.desired.description,
+                            permission_sets_id_map[
+                                role_diff_pair.desired.permission_set_name
+                            ],
+                            access_scope_id_map[
+                                role_diff_pair.desired.access_scope.name
+                            ],
+                        )
+                    except Exception as e:
+                        logging.error(
+                            "Failed to update role: %s\n\t%s",
+                            role_diff_pair.desired.name,
+                            e,
+                        )
+                        err_count += 1
+                        continue
+                logging.info("Updated role: %s", role_diff_pair.desired.name)
+        return err_count
+
+    def reconcile(
+        self,
+        desired: list[AcsRole],
+        current: list[AcsRole],
+        acs: AcsApi,
+        auth_provider_id: str,
+        dry_run: bool,
+    ) -> int:
+        err_count = 0
+        diff = diff_iterables(current, desired, lambda x: x.name)
+        if len(diff.add) > 0:
+            err_count += self.add_rbac(diff.add, acs, auth_provider_id, dry_run)
+        if len(diff.delete) > 0:
+            err_count += self.delete_rbac(diff.delete, acs, auth_provider_id, dry_run)
+        if len(diff.change) > 0:
+            err_count += self.update_rbac(diff.change, acs, auth_provider_id, dry_run)
+        return err_count
+
+    def run(
+        self,
+        dry_run: bool,
+    ) -> None:
+        gqlapi = gql.get_api()
+        instance = self.get_acs_instance(gqlapi.query)
+
+        vault_settings = get_app_interface_vault_settings()
+        secret_reader = create_secret_reader(use_vault=vault_settings.vault)
+        token = secret_reader.read_all_secret(instance.credentials)
+
+        desired = self.get_desired_state(gqlapi.query)
+
+        errors_occurred = 0
+        with AcsApi(
+            instance={"url": instance.url, "token": token[instance.credentials.field]}
+        ) as acs_api:
+            current = self.get_current_state(acs_api, instance.auth_provider.q_id)
+            errors_occurred = self.reconcile(
+                desired, current, acs_api, instance.auth_provider.q_id, dry_run
+            )
+
+        if errors_occurred:
+            sys.exit(ExitCodes.ERROR)
```

Comment:
For new integrations, we don't use `sys.exit` anymore, just throw an exception like `RuntimeError` or any other custom exception. You can also include the actual number of errors in the error message too.

That `errors_occurred = 0` is redundant as the value is always the return of `self.reconcile`. 

`err_count += 1` always have a following `continue`, if you want to remove that duplication, can extract process single role to a function, directly throw the exception, catch it like this:

```python
        for role in to_add.values():
            try:
                add_rbac_for_role(role, ...)
            except Exception as e:
                logging.error(...)
                err_count += 1
```

If you want to try the new feature of `ExceptionGroup` to collect all exceptions, you can use it to replace custom `error_count`, like this

```python
        errors = []
        for role in to_add.values():
            try:
                add_rbac_for_role(role, ...)
            except Exception as e:
                errors.append(e)
        if errors:
            raise ExceptionGroup("Add role errors", errors)
```

Or you can choose throw it at a higher level in `reconcile`

```python
        errors = []
        diff = diff_iterables(current, desired, lambda x: x.name)
        if diff.add:
            errors.extend(self.add_rbac(diff.add, acs, auth_provider_id, dry_run))
        if diff.delete:
            errors.extend(self.delete_rbac(diff.delete, acs, auth_provider_id, dry_run))
        if diff.change:
            errors.extend(self.update_rbac(diff.change, acs, auth_provider_id, dry_run))
        if errors:
            raise ExceptionGroup("Reconcile errors occurred", errors)
```

That will make `run` logic clean as it doesn't need to care about error handling anymore.

If the error message in each function call like `acs.create_access_scope` is good enough, then you don't need to catch and rethrow, just write success path handling. You may find the new extracted function can further improve readability as it only need related value passed in like `access_scope_id_map[DEFAULT_ADMIN_SCOPE_NAME]` and `permission_sets_id_map[role.permission_set_name]`.


Btw I just noticed that `acs.get_access_scopes()`, `acs.get_permission_sets()` and `get_groups()` called multiple times, you may want to call it once and pass to later usages. I don't know the relationship between `get_access_scope_by_id` and `get_access_scopes`, will `get_access_scopes` return all data without pagination? `get_access_scope_by_id` is called in a loop to fetch one by one, can it be covered by `get_access_scopes`?

Reply:
Regarding the usage of `AcsApi` calls to obtain existing state, you're right, this could be reduced significantly.

[refactored](https://github.com/app-sre/qontract-reconcile/pull/3899/commits/634292c48c1788dd0a51cceb8aa350a5cda94c05)

Reply:
Thanks for the suggestions for error handling. Much cleaner approach.

 [error handling refactor](https://github.com/app-sre/qontract-reconcile/pull/3899/commits/93d7313e6c8135b75c6af2c5f431de941abcfe7a)